### DISCUSSION TITLE
Delegate a custom event from .js into c# context

### DISCUSSION THREAD

**Discussion by vreljanski** [hardkoded/puppeteer-sharp#2061] [Q&A]
I managed to inject a event listener function into .js page i red. Now how would I make a delegate in C# context and handle that function call as an event?

**Comment by kblok** [comment]
You could try using [ExposeFunctionAsync](https://github.com/hardkoded/puppeteer-sharp/blob/master/lib/PuppeteerSharp.Tests/PageTests/ExposeFunctionTests.cs#L22).

**Reply by vreljanski** [reply]
Something like this is the idea, just... won`t bulge...

// Define the function you want to handle as an event
string function = @"
    function handleClick(event) {
        console.log('Click event:', event);
    }
";

// Evaluate the function and bind it to the 'click' event
await page.EvaluateExpressionAsync($"document.addEventListener('click', {function})");

// Inject a script that will listen for the 'click' event and raise it in the C# context
page.Frame.AddScriptToEvaluateOnNewDocument(@"
    window.addEventListener('click', event => {
        window.CSharp.raiseEvent('click', event);
    });
");

// Subscribe to the 'click' event in the C# context
page.Event += (sender, e) => {
    if (e.Name == "click") {
        Console.WriteLine("Click event received in C# context:", e.Args[0]);
    }
};

**Comment by vreljanski** [comment]
I am actually injecting this on particular set of div's

  //add listener!
  await divElement.EvaluateFunctionAsync(@"div => {
  div.addEventListener('DOMSubtreeModified', event => console.log(event.target.innerHTML));
  }");
  
  Now... How to fire event in C# context each time this event => console.log is fired... thats practically it.

**Reply by kblok** [reply]
[This is playwright code](https://github.com/microsoft/playwright-dotnet/issues/1377#issuecomment-854812270). But it's pretty much the same.

**Reply by vreljanski** [reply]
This seems like it could bulge! I will go reset now 11 str8h of coding. Puppeteer is nice and fast. I am literaly "riding" the instantiated target spa and reading as it happens. Now just to drag it to c# context.

Thanks, this looks like something that could save the day tomorrow after refresh.
